hadoop-hdds/interface-server/src/main/proto/ScmServerProtocol.proto

```diff
@@ -230,6 +233,60 @@ message SortDatanodesResponseProto{
   repeated DatanodeDetailsProto node = 1;
 }
 
+message NodeImpl {
+  required string name = 1;
+  required string location = 2;
+  required uint32 cost = 3;
+}
+
+message CommandCount {
+  required SCMCommandProto.Type command = 1;
+  required uint32 count = 2;
+}
+
+message NodeStatus {
+  required NodeOperationalState operationalState = 1;
+  required NodeState health = 2;
+  required uint64 opStateExpiryEpochSeconds = 3;
+}
+
+message DatanodeInfo {
+  required DatanodeDetailsProto datanodeDetails = 1;
+  required uint64 lastHeartbeatTime = 2;
+  required uint64 lastStatsUpdatedTime = 3;
+  required uint32 failedVolumeCount = 4;
+  repeated StorageReportProto storageReportProto = 5;
+  repeated MetadataStorageReportProto metadataStorageReportProto = 6;
+  required LayoutVersionProto layoutVersionProto = 7;
+  repeated CommandCount commandCount = 8;
+  required NodeStatus nodeStatus = 9;
+}
+
+message NodeType {
+  optional DatanodeDetailsProto datanodeDetails = 1;
+  optional DatanodeInfo datanodeInfo = 2;
```

Comment:
Doesn't `DatanodeInfo` also include `DatanodeDetailsProto`

Reply:
Yes, that's correct! I have included `DatanodeDetailsProto` under the `DatanodeInfo` proto definition:

```
message DatanodeInfo {
  required DatanodeDetailsProto datanodeDetails = 1;
  required uint64 lastHeartbeatTime = 2;
  required uint64 lastStatsUpdatedTime = 3;
  required uint32 failedVolumeCount = 4;
  repeated StorageReportProto storageReportProto = 5;
  repeated MetadataStorageReportProto metadataStorageReportProto = 6;
  required LayoutVersionProto layoutVersionProto = 7;
  repeated CommandCount commandCount = 8;
  required NodeStatus nodeStatus = 9;
}
```

Reply:
I think the question was intended to highlight that `NodeType` has `DatanodeDetailsProto` twice: one directly, one inside `DatanodeInfo`.

Reply:
I apologize if I have not understood the question initially, I was trying to serialize and deserialize the interface: `Node` from the package: `org.apache.hadoop.hdds.scm.net` (while interfaces in general are not really meant to be serialized/deserialized).
- `Node` has a couple of implementations (direct/indirect): `DatanodeDetails`, `DatanodeInfo`, `InnerNode`, `InnerNodeImpl`, `NodeImpl`.
- Hence, the `DatanodeDetailsProto` proto object has been included inside `NodeType` (which is the corresponding proto object for the `Node` interface).
- But again, the `DatanodeInfo` class extends `DatanodeDetails` due to which the `DatanodeDetailsProto` proto object has been added once more inside the `DatanodeInfo` proto object.

Please suggest if there is a better way to approach this, thanks!

Reply:
Currently, only DatanodeDetails is used in the childrenMap. So you can try to simplify the new protos as 

```
message InnerNode {
  required NodeImpl node = 1;
  required uint32 numOfLeaves = 2;
  repeated ChildrenInfo children = 3;
}

message NodeType {
  optional DatanodeDetailsProto datanodeDetails = 1;
}

message ChildrenInfo {
  required string networkName = 1;
  required NodeType nodeType = 2;
}

```

BTW,  you missed level and parent in NodeImpl, which are both very important, especially parent.  The topology tree is bi-direction linked. Without the parent link, the distance calculation cannot be done.

Reply:
![image (12)-2 (1)](https://github.com/apache/ozone/assets/46785609/fae8bd65-9c57-475a-ba2a-c2c6631d5a50)

- While debugging I also happened to see `childrenMap`'s value be of type `InnerNodeImpl` apart from being `DatanodeDetails` (please find the image), I will try including both these types under the `NodeType` proto object.

```
message NodeType {
  optional DatanodeDetailsProto datanodeDetails = 1;
  optional InnerNode innerNode = 2;
}
```

Reply:
![image (13)-2 (1)](https://github.com/apache/ozone/assets/46785609/418e744e-b00d-404b-9f73-39ee6b61c8e7)
- I was facing a few challenges while trying to include the parent information during serialization/deserialization. Every child object has a reference to its parent and similarly the parent has a reference to its child (as shown in the image: parent shown in white, child shown in yellow, for example). This sort of makes the serialization/deserialization go on infinitely deeper. 
- I was looking at the existing implementation of serialization/deserialization of `DatanodeDetails` (which extends `NodeImpl`), where we are currently not including the parent information.
- I went with the second constructor of `NodeImpl` to create the object (which does not include the parent). 
- I am trying to check if there's a better way to approach this, open to any suggestions!

Reply:
Handled the case for including the parent information as part of `ScmBlockLocationProtocolClientSideTranslatorPB#setParent` call.

Reply:
I have simplified the proto object definitions as suggested and updated the patch, thanks!